design/20230313_tink_worker_failure_reason.md

```diff
@@ -0,0 +1,28 @@
+# Tink Worker Failure Reasons
+
+## Context
+
+Tink Worker is the client launched by an Operating System Installation Environment (OSIE) that communicates with Tink Server to retrieve actions to run on the node. Actions are OCI images that Tink Worker can launch using a container runtime such as Docker. In the event an action fails, Tinkerbell does not provide mechanisms for the user defined action to communicate why it failed. This makes debugging `Workflow`s difficult as users leveraging `kubectl` cannot observe a specific failure reason and must resort to running commands directly on the node.
+
+As part of the Tink CRD Refactor proposal we introduced `Reason` and `Message` fields to indiciate why an action entered a failure state. The proposal does not detail how these fields are populated but envisages at minimum the `Reason` being used to communicate timeout failures. 
+
+* A `Reason` is a machine readable CamelCase word or phrase that succinctly describes the failure reason. 
+* A `Message` is a human readable string that elaborates on the failure reason to provide specifics.
+
+This proposal lays out a contract for action containers to communicate why it exited with a non-zero exit code.
+
+## Goals/Non-goals
+
+**Goals**
+
+- Define a contract for action images to communicate failure information that is exposed via associated custom resource definitions and is therefore inspectable with `kubectl`.
+
+## Proposal
+
+Actions can communicate a failure reason by writing to `/dev/reason`.
+
+Actions can communicate a failure message by writing to `/dev/message`.
```

Comment:
As these files aren't devices, what do you think about choosing a different location instead of `/dev`?

Reply:
I'm open to changing the location. 

What I'm going for is for the backend reading from these files to be transparent to the user. I.e. they might not be literal text files.

Reply:
Resolved based on the community meeting discussion: use `/tinkerbell/failure-reason` and `/tinkerbell/failure-message`.